### ISSUE TITLE
problem with train_recognizer.py

### ISSUE THREAD

**Issue by jiawensong** [dmlc/gluon-cv#1065]
in the script ,trainer use kvstore but set update_on_kvstore=false.Have someone know what this mean?

**Comment by bryanyzhu** [comment]
Hi, I assume you mean the `train_recognizer.py` in action recognition application, which is this [file](https://github.com/dmlc/gluon-cv/blob/master/scripts/action-recognition/train_recognizer.py). 

`kvstore` is used when you want to do distributed training. KVStore server supports two modes, one which aggregates the gradients and updates the weights using those gradients (update_on_kvstore=True), and second where the server only aggregates gradients (update_on_kvstore=False). In the latter case, when a worker process pulls from kvstore, it gets the aggregated gradients. The worker then uses these gradients and applies the weights locally.

We choose the second mode, `update_on_kvstore=False` because we can easily change learning rate on the worker side.

For more information, you can read about distributed training [here](https://mxnet.apache.org/api/faq/distributed_training).

**Comment by jiawensong** [comment]
ok, thanks for your explain. I have got the reason now.

**Comment by jiawensong** [comment]
another question. Is this setting have some advantages, like faster training speed or less limited by band width?

**Comment by bryanyzhu** [comment]
Sorry for the slow response, yes, you are right. Updating the parameters on local machines will have faster speed. And since we only communicate gradients, it is also less limited by bandwidth.

**Comment by jiawensong** [comment]
> Sorry for the slow response, yes, you are right. Updating the parameters on local machines will have faster speed. And since we only communicate gradients, it is also less limited by bandwidth.

Can you simply describe the way to update weights?In my understanding, if weights are updated in parameter server, we just need to send weights to each worker. And the size of weights may have little difference to gradients, so why setting update_on_kvstore=false sends less data?

**Comment by bryanyzhu** [comment]
Sorry, my previous reply is not accurate. They send the same data, so the bandwidth requirement is the same.

**Comment by bryanyzhu** [comment]
Any other problem with `train_recognizer.py`? If not, I will close the issue.

**Comment by jiawensong** [comment]
> Any other problem with `train_recognizer.py`? If not, I will close the issue.

Thanks for your reply. Exactly I am still confused about the reason of this setting. Update on local machine means local machine don't need to wait for update weights in server node, but in my understanding update weitghts costs little time and this setting will lead to computing resources wasted.

**Comment by bryanyzhu** [comment]
As for speed, it depends. 

When update_on_kvstore is True, which means update weights in the server. As you said, updating weights usually cost little time so it may save some time. 

However, updating weights on the server node is done on CPU. So if you have a strong CPU with many cores in the server node, then yes, setting update_on_kvstore to True is better. But If you have a weak CPU or you are doing lots of CPU work on the server node, it may slow down the training. 

In my case, as video dataloader is extremely CPU intensive, I decide to set update_on_kvstore to False. But I didn't compare these two settings, so I actually don't know which is better. 

@eric-haibin-lin  is an expert in distributed training. @eric-haibin-lin  Please correct me or leave more comments. Thank you.

**Comment by eric-haibin-lin** [comment]
In terms of speech, most of the time I don't see much difference between the two. However, in terms of usability, `update_on_kvstore=False` is more flexible because the worker can print the learning rate used for each batch.  

> update weitghts costs little time and this setting will lead to computing resources wasted.

Yes, kind of. One thing to note is that CPU is slow at handling float16 arithmetics. If you have casted your network to fp16 to do mixed precision training with fp16 gradients, `update_on_kvstore=True` may create a bottleneck at the server side. Although it is not used in `train_recognizer` script.

**Comment by jiawensong** [comment]
Thanks for both of your replies. I think the answers are clear enough to explain the reason of this setting in train_recognizer.
